### PR TITLE
Add decimal average function

### PR DESCRIPTION
**Pull request by karteekmurthys**
Decimal Average function returns a DECIMAL type as output and the precision and scale match input columns precision/scale.

Today, the `AverageAggregate`  class expects the Intermediate/Accumulator and Final aggregation types to be DOUBLE or REAL. But for this class to handle DECIMAL types, the accumulator and final aggregations needs to be `SHORT_DECIMAL or LONG_DECIMALS`. This PR refactors existing `AverageAggregate` class to be a template function in-order to extend the class to handle DECIMALS as well.

|     TYPE                     |      Partial Aggregate type   |    Final Aggregate type  |
----------------------|---------------------------|------------------------- |
SHORT_DECIMAL(a_precision, a_scale)       |          LONG_DECIMAL(38, a_scale)       |       SHORT_DECIMAL(a_precision, a_scale)     |
LONG_DECIMAL(a_precision, a_scale)       |          LONG_DECIMAL(38, a_scale)       |       LONG_DECIMAL(a_precision, a_scale)     |

### CODE REVIEW

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 1]
```diff
@@ -24,30 +24,31 @@ namespace facebook::velox::aggregate {
...
 struct SumCount {
-  double sum{0};
+  TSumType sum{0};
   int64_t count{0};
 };
 
 // Partial aggregation produces a pair of sum and count.
 // Final aggregation takes a pair of sum and count and returns a real for real
 // input types and double for other input types.
 // T is the input type for partial aggregation. Not used for final aggregation.
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:36]: update comments.

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 2]
```diff
@@ -24,30 +24,31 @@ namespace facebook::velox::aggregate {
...
-template <typename T>
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:36]: `Tintermediate` -> `TAccumulator` to match AggregateSum?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 3]
```diff
@@ -274,19 +283,31 @@ class AverageAggregate : public exec::Aggregate {
       } else {
         clearNull(rawNulls, i);
         auto* sumCount = accumulator(group);
-        rawValues[i] = (TResult)sumCount->sum / sumCount->count;
+        if constexpr (
+            (std::is_same_v<TResult, double> || std::is_same_v<TResult, float>)&&(
+                std::is_same_v<TIntermediate, double> ||
+                std::is_same_v<TIntermediate, float>)) {
+          rawValues[i] = (TResult)sumCount->sum / sumCount->count;
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:290]: `TResult(sumcCount->sum)`

**Review by majetideepak — COMMENTED** [review]
few comments. As discussed offline, let's ensure the rounding, as well as overflow behavior, matches Presto.

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 4]
```diff
@@ -279,19 +309,98 @@ class AverageAggregate : public exec::Aggregate {
...
+    exec::Aggregate::clearNull(group);
+  }
+  accumulator(group)->sum =
+      checkedPlus<int128_t>(accumulator(group)->sum, value.unscaledValue());
+  accumulator(group)->count += 1;
+}
 
 void checkSumCountRowType(TypePtr type, const std::string& errorMessage) {
-  VELOX_CHECK_EQ(type->kind(), TypeKind::ROW, "{}", errorMessage);
-  VELOX_CHECK_EQ(
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:293]: Why remove this check?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 5]
```diff
@@ -279,19 +309,98 @@ class AverageAggregate : public exec::Aggregate {
...
-      type->childAt(0)->kind(), TypeKind::DOUBLE, "{}", errorMessage);
+  VELOX_CHECK(
+      type->kind() == TypeKind::ROW || type->kind() == TypeKind::LONG_DECIMAL,
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:401]: Why do we need only the LONG_DECIMAL check in `checkSumCountRowType`
Can you add a comment?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 6]
```diff
@@ -279,19 +309,98 @@ class AverageAggregate : public exec::Aggregate {
...
+inline int32_t
+AverageAggregate<UnscaledLongDecimal, int128_t, UnscaledShortDecimal>::
+    accumulatorAlignmentSize() const {
+  return static_cast<int32_t>(sizeof(int128_t));
+}
+
+template <>
+template <bool tableHasNulls>
+void AverageAggregate<UnscaledShortDecimal, int128_t, UnscaledShortDecimal>::
+    updateNonNullValue(char* group, UnscaledShortDecimal value) {
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:366]: The body of the three specializations is the same. Probably use a single template function.

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 7]
```diff
@@ -279,19 +309,98 @@ class AverageAggregate : public exec::Aggregate {
       } else {
         clearNull(rawNulls, i);
         auto* sumCount = accumulator(group);
-        rawValues[i] = TResult(sumCount->sum) / sumCount->count;
+        if constexpr (
+            (std::is_same_v<TResult, double> || std::is_same_v<TResult, float>)&&(
+                std::is_same_v<TAccumulator, double> ||
+                std::is_same_v<TAccumulator, float>)) {
+          rawValues[i] = TResult(sumCount->sum) / sumCount->count;
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:316]: Can simply be `if constexpr (decimal types) {} else {}` ?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 8]
```diff
@@ -235,7 +260,12 @@ class AverageAggregate : public exec::Aggregate {
       rows.applyToSelected([&](vector_size_t i) {
         auto decodedIndex = decodedPartial_.index(i);
         totalCount += baseCountVector->valueAt(decodedIndex);
-        totalSum += baseSumVector->valueAt(decodedIndex);
+        if (std::is_same_v<TAccumulator, int128_t>) {
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:263]: if constexpr?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 9]
```diff
@@ -132,26 +135,48 @@ class AverageAggregate : public exec::Aggregate {
...
+      if constexpr (
+          std::is_same_v<TInput, UnscaledShortDecimal> ||
+          std::is_same_v<TInput, UnscaledLongDecimal>) {
+        rows.applyToSelected([&](vector_size_t i) {
+          totalSum = checkedPlus<int128_t>(
+              totalSum, decodedRaw_.valueAt<TInput>(i).unscaledValue());
+        });
+      } else {
+        rows.applyToSelected([&](vector_size_t i) {
+          totalSum += decodedRaw_.valueAt<TInput>(i);
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:177]: Shouldn't we be using checkedPlus here as well?

Comment (karteekmurthys) [reply velox/functions/prestosql/aggregates/AverageAggregate.cpp:177]: This is for double as intermediate type and there is no chekedPlus overloading for that.

File: velox/exec/tests/utils/QueryAssertions.cpp [hunk 10]
```diff
@@ -438,6 +438,9 @@ velox::variant rowVariantAt(const VectorPtr& vector, vector_size_t row) {
       values.push_back(arrayVariantAt(child, row));
     } else if (child->typeKind() == TypeKind::MAP) {
       values.push_back(mapVariantAt(child, row));
+    } else if (child->typeKind() == TypeKind::LONG_DECIMAL) {
```
Comment (majetideepak) [inline velox/exec/tests/utils/QueryAssertions.cpp:441]: Why don't we need a check for `SHORT_DECIMAL`?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp [hunk 11]
```diff
@@ -37,6 +37,8 @@ struct SumCount {
 //   ------------------|---------------------|------------------
 //     REAL            |     DOUBLE          |    REAL
 //     ALL INTs        |     DOUBLE          |    DOUBLE
+//     SHORT_DECIMAL   | LONG_DECIMAL        | SHORT_DECIMAL
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/AverageAggregate.cpp:40]: nit: align the entries in the second and third column

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp (hunk 7 above)
Comment (majetideepak) [reply velox/functions/prestosql/aggregates/AverageAggregate.cpp:288]: can you address this comment?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp (hunk 5 above)
Comment (majetideepak) [reply velox/functions/prestosql/aggregates/AverageAggregate.cpp:326]: can you address this comment?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp (hunk 4 above)
Comment (majetideepak) [reply velox/functions/prestosql/aggregates/AverageAggregate.cpp:293]: can you address this comment?

File: velox/functions/prestosql/aggregates/tests/AverageAggregationTest.cpp [hunk 12]
```diff
@@ -199,5 +203,67 @@ TEST_F(AverageAggregationTest, partialResults) {
...
+TEST_F(AverageAggregationTest, avgDecimal) {
+  // facebook::velox::exec::test::AggregateTypeResolver resolver(step);
+  auto runAndCompare = [&](const std::string& exprStr,
+                           std::vector<VectorPtr> input,
+                           VectorPtr expectedResult,
+                           bool isSingle = true) {
+    // Need to use PlanBuilder as it registers a AggregateTypeResolver Vs
+    // evaluate which would compile the expressions through Simple/Vector
+    // function resolvers.
+    PlanBuilder builder;
```
Comment (majetideepak) [inline velox/functions/prestosql/aggregates/tests/AverageAggregationTest.cpp:216]: Can you add a comment as to why we cannot use `testAggregations`?

File: velox/functions/prestosql/aggregates/AverageAggregate.cpp (hunk 5 above)
Comment (karteekmurthys) [reply velox/functions/prestosql/aggregates/AverageAggregate.cpp:326]: This is a mistake, the check must be split. Fixed it.

**Comment by yzhang1991** [comment]
@karteekmurthys Please comment and close.

**Comment by karteekmurthys** [comment]
Closing this. An approach similar to JAVA has been implemented here: https://github.com/facebookincubator/velox/pull/2883

**Pull request closed without merging**